api/stages_api.py

```diff
@@ -0,0 +1,251 @@
+      self._add_given_stage_vals(stage, kwargs, self.SHIPPING_FIELDS)
+      gate_type = core_enums.GATE_SHIP
+
+    stage.put()
+
+    # If we should create a gate and this is a stage that requires a gate,
+    # create it.
+    if create_gate and gate_type is not None:
+      self._create_gate_for_stage(
+          stage.feature_id, stage.key.integer_id(), gate_type)
+
+
+  def do_get(self, **kwargs):
+    """Return a specified stage based on the given ID."""
+    stage_id = kwargs.get('stage_id', None)
+
+    if stage_id is None:
+      self.abort(404, msg='No stage specified.')
+    
+    stage: Stage | None = Stage.get_by_id(stage_id)
+    if stage is None:
+      self.abort(404, msg=f'Stage {stage_id} not found')
+
+    return self._stage_to_json_dict(stage)
+
+  def do_post(self, **kwargs):
+    """Create a new stage."""
+    feature_id = kwargs.get('feature_id', None)
+    if feature_id is None:
+      self.abort(404, msg='No feature specified.')
+
+    feature: FeatureEntry | None = FeatureEntry.get_by_id(feature_id)
+    if feature is None:
+      self.abort(404, msg=f'Feature {feature_id} not found')
+
+    # Validate the user has edit permissions and redirect if needed.
+    redirect_resp = permissions.validate_feature_edit_permission(
+        self, feature_id)
+    if redirect_resp:
+      return redirect_resp
+
+    if 'stage_type' not in kwargs:
+      self.abort(404, msg='Stage type not specified.')
+
+    # Create the stage.
+    stage = Stage(feature_id=feature_id)
+    # Add the specified field values to the stage. Create a gate if needed.
+    self._update_stage_vals(
+        stage, feature.feature_type, kwargs, create_gate=True)
+
+    # Return  the newly created stage ID.
+    return {'message': 'Stage created.', 'stage_id': stage.key.integer_id()}
+  
+  def do_patch(self, **kwargs):
+    """Update an existing stage based on the stage ID."""
+    stage_id = kwargs.get('stage_id', None)
+
+    if stage_id is None:
+      self.abort(404, msg='No stage specified.')
+
+    stage: Stage | None = Stage.get_by_id(stage_id)
+    if stage is None:
+      self.abort(404, msg=f'Stage {stage_id} not found')
+
+    feature: FeatureEntry | None = FeatureEntry.get_by_id(stage.feature_id)
+    if feature is None:
+      self.abort(404, msg=(f'Feature {stage.feature_id} not found '
+          f'associated with stage {stage_id}'))
+    feature_id = feature.key.integer_id()
+    # Validate the user has edit permissions and redirect if needed.
+    redirect_resp = permissions.validate_feature_edit_permission(
+        self, feature_id)
+    if redirect_resp:
+      return redirect_resp
+
+    # The stage_type field should not be able to change.
+    if 'stage_type' in kwargs:
+      self.abort(404, msg='stage_type field cannot be mutated.')
+
+    # Update specified fields. No need to create a gate for existing stage.
+    self._update_stage_vals(
+        stage, feature.feature_type, kwargs, create_gate=False)
```

Comment:
The kwargs here is just the URL path parameters parsed by Flask.  I think you want the JSON content in the body of the POST request.  You can get that with:
`json_body = self.request.get_json(force=True, silent=True) or {}`

Because that will be needed in a couple of different places, you could define a one-line method `get_json_param_dict()` in BaseHandler.

Reply:
Yeah, my approach was wrong here and needed a good refactoring, but it should be correct now. 🙂